deployment/7_upgrade_v2.js

```diff
@@ -0,0 +1,133 @@
+const BACKWARD_COMPATIBILITY = 1;
+
+const deploy = async (network) => {
+
+    const newModuleWrappers = [];
+    const newVersion = {};
+
+    ////////////////////////////////////
+    // Setup
+    ////////////////////////////////////
+
+    const manager = new DeployManager(network);
+	await manager.setup();
+
+	const configurator = manager.configurator;
+	const deployer = manager.deployer;
+    const abiUploader = manager.abiUploader;
+    const versionUploader = manager.versionUploader;
+    const deploymentWallet = deployer.signer;
+    const config = configurator.config;
+
+    const ModuleRegistryWrapper = await deployer.wrapDeployedContract(ModuleRegistry, config.contracts.ModuleRegistry);
+    const MultiSigWrapper = await deployer.wrapDeployedContract(MultiSig, config.contracts.MultiSigWallet);
+    const multisigExecutor = new MultisigExecutor(MultiSigWrapper, deploymentWallet, config.multisig.autosign);
+    
+    console.log('Config:', config);
+
+    ////////////////////////////////////
+    // Deploy new modules
+    ////////////////////////////////////
+
+    const NftTransferWrapper = await deployer.deploy(
+        NftTransfer,
+        {},
+        config.contracts.ModuleRegistry,
+        config.modules.GuardianStorage
+    );
+    newModuleWrappers.push(NftTransferWrapper);  
+    
+    ///////////////////////////////////////////////////
+    // Update config and Upload new module ABIs
+    ///////////////////////////////////////////////////
+
+    configurator.updateModuleAddresses({
+        NftTransfer: NftTransferWrapper.contractAddress
+    });
+
+    const gitHash = require('child_process').execSync('git rev-parse HEAD').toString('utf8').replace(/\n$/, '');
+    configurator.updateGitHash(gitHash);
+    await configurator.save();
+
+    await Promise.all([
+        abiUploader.upload(NftTransferWrapper, "modules")
+    ]);
+
+    ////////////////////////////////////
+    // Register new modules
+    ////////////////////////////////////
+
+    for (let idx = 0; idx < newModuleWrappers.length; idx++) {
+        let wrapper = newModuleWrappers[idx];
+        await multisigExecutor.executeCall(ModuleRegistryWrapper, "registerModule", [wrapper.contractAddress, utils.asciiToBytes32(wrapper._contract.contractName)]);
+    }
+
+    ////////////////////////////////////
+    // Deploy and Register upgraders
+    ////////////////////////////////////
+
+    const toAdd = newModuleWrappers.map((wrapper) => {
+        return {
+            'name': wrapper._contract.contractName,
+            'address': wrapper.contractAddress
+        };
+    });
+    let fingerprint; 
+
+    const versions = await versionUploader.load(BACKWARD_COMPATIBILITY);
+    for (let idx = 0; idx < versions.length; idx++) {
+        const version = versions[idx];
+        const moduleNames = MODULES_TO_DISABLE.concat(MODULES_TO_ENABLE);
+        const toRemove = version.modules.filter(module => moduleNames.includes(module.name));
```

Comment:
shouldn't we use `MODULES_TO_DISABLE` instead of `moduleNames` here? (even though it makes not difference for that particular upgrade)

Reply:
No because we also need to remove the modules that we are adding (if they are present). 

Say we are in `Test` and the target version adds the NftModule. We go from 1.0.x to 1.1.0. We then realise that there is a bug in the NftModule so we update it and make a patch to go to 1.1.1. Wallets that did not upgrade to 1.1.0 will need to add the new version of the NftModule, while wallets that upgraded to 1.1.0 need to remove their current NftModule then add the new version. The current script takes care of these 2 scenarios by concatenating `MODULES_TO_DISABLE` with `MODULES_TO_ENABLE`.

Reply:
ok